core.html

```diff
@@ -42,70 +34,90 @@ <h2>Method Specific Identifier</h2>
 
     <p>The peer DID scheme is defined by the following ABNF (see [[RFC5234]] for syntax):</p>
     <pre class="example" class="ABNF" title="ABNF for peer DIDs">
-peer-did = "did:peer:" numalgo encalgo "-" numbasis
+peer-did = "did:peer:" numalgo transform encnumbasis
```

Comment:
Hmm, not quite, but close... what about this instead:

> peer-did = "did:peer:" numalgo ":" multibase_identifier multicodec_identifier encnumbasis

For an ed25519 key, that would be:

> did:peer:1:z 0xed ...32-bytes of public key...

So, a realistic key would look like this:

> did:peer:1:z3ELxxhw7Etnj2phqrwTWx3Uu

Yes, it's two extra bytes (one for the colon, another for the multicodec_identifier), but that would enable to write super simple did:peer/did:key public key parsers... thoughts?

Reply:
The thing that's being encoded in encnumbasis is NOT a key; it's the sha256 hash of the genesis doc. The genesis doc can have multiple keys in it, and this DID method allows any key type. So we don't want 0xED, we want 0x12 (which is the multicodec identifier for a sha256 hash). And the value that comes after is 32 bytes, whereas the one you're showing is 16 bytes.

Reply:
I don't understand how you are imagining the multicodec identifier (0xed in your example) fits into the ABNF. There is no 0xed in your realistic key; you are encoding it in base58, right? So if that's the case, then you can't directly represent multicodec in the ABNF; you have to do what I did and use encnumbasis, which includes the multicodec and the data that the multicodec prefix describes -- both transformed according to the multibase. Am I missing something?

Reply:
@msporny: You may want to look at the actual graphic that goes with this text, because I think it explains what is happening more clearly than words can: https://dhh1128.github.io/peer-did-method-spec/composition.png

Reply:
> The thing that's being encoded in encnumbasis is NOT a key; it's the sha256 hash of the genesis doc. 

Ah, I see... this is a fundamental difference between did:peer and did:key, then... with no smooth upgrade path, as I had imagined. I *thought* at some point we had pondered starting did:peer in the same way that did:key was started... that is, both can start with a simple DID Document that's generated purely from a public key.

If that's not the case, then there is almost no overlap between the did:peer and did:key methods... but there is still benefit in standardizing on multibase/multicodec.

> The genesis doc can have multiple keys in it, and this DID method allows any key type. So we don't want 0xED, we want 0x12 (which is the multicodec identifier for a sha256 hash). 

Yes, now we're on the same page... and you do want 0x12...

> And the value that comes after is 32 bytes, whereas the one you're showing is 16 bytes.

Right again.

> And the value that comes after is 32 bytes, whereas the one you're showing is 16 bytes.

Reply:
> You may want to look at the actual graphic that goes with this text

Yes, that was helpful. The only thing you're missing is the 0x20 (hash output length), which goes directly after the 0x12 (sha2-256).

Reply:
I've added the 0x20 for hash output length.

I will address the possibility of starting a peer did with a naked key (the issue @msporny raised earlier in this comment stream) in a separate PR. In this PR, the spec still doesn't allow that.